Custom printing for KFuzz

gtrepta opened the pull request:
This implements a new class `KFuzzTestPrinter` which allows custom printing routines to be called during the fuzzing of a property test in K for better output to the user.

gtrepta committed:
Fix hypothesis_args not being passed from KFuzz class

gtrepta committed:
Allow passing check to llvm_interpret_raw

gtrepta committed:
Create KFuzzTestPrinter, for custom output during fuzz testing

tothtamas28 commented on pyk/src/pyk/ktool/krun.py:
```diff
@@ -424,19 +424,22 @@ def llvm_interpret(definition_dir: str | Path, pattern: Pattern, *, depth: int |
-def llvm_interpret_raw(definition_dir: str | Path, kore: str, depth: int | None = None) -> CompletedProcess:
+def llvm_interpret_raw(
+    definition_dir: str | Path, kore: str, depth: int | None = None, check: bool = True
+) -> CompletedProcess:
```
```suggestion
def llvm_interpret_raw(
    definition_dir: str | Path, kore: str, *, depth: int | None = None, check: bool = True
) -> CompletedProcess:
```

tothtamas28 commented on pyk/src/pyk/ktool/krun.py:
```diff
@@ -424,19 +424,22 @@ def llvm_interpret(definition_dir: str | Path, pattern: Pattern, *, depth: int |
 
     Raises:
-        CalledProcessError: If the interpreter fails.
+        CalledProcessError: If check=True and the interpreter fails.
```
```suggestion
        CalledProcessError: If ``check`` and the interpreter fails.
```

tothtamas28 commented on pyk/src/pyk/ktool/kfuzz.py:
```diff
@@ -20,13 +20,32 @@
 
+class KFuzzTestPrinter:
+    def print_test_prologue(self) -> None:
+        pass
```
Consider

```suggestion
    def print_test_prologue(self) -> str:
        pass
```

for all methods.

gtrepta replied:
I actually would prefer this stays `None` and allow more flexibility. Primarily this would be used to report the status of the fuzzing to the user, but as far as how that's done, either with `print` or `logging` or some other creative way should be up to the application.

Given that, I've renamed it from `KFuzzTestPrinter` to `KFuzzTestCaseHandler`. That's a bit of a long name, open to suggestions.

tothtamas28 commented on pyk/src/pyk/ktool/kfuzz.py:
```diff
@@ -20,13 +20,32 @@
 
+class KFuzzTestPrinter:
+    def print_test_prologue(self) -> None:
+        pass
```
```suggestion
class KFuzzTestPrinter(ABC):
    def print_test_prologue(self) -> None: ...
```

gtrepta committed:
Make KFuzzTestPrinter an abstract class.

Also remove the prologue/epilogue, as the app can just do those itself
outside of calling KFuzz

gtrepta committed:
Formatting/Documentation

gtrepta committed:
Propogate check parameter up to llvm_interpret

gtrepta committed:
Rename Printer->TestCaseHandler

gtrepta committed:
formatting

tothtamas28 commented on pyk/src/pyk/ktool/kfuzz.py:
```diff
@@ -20,13 +21,34 @@
     from ..kore.syntax import Pattern
 
 
+class KFuzzTestCaseHandler(ABC):
```
Maybe `KFuzzHandler` or `TestCaseHandler`?

tothtamas28 replied:
Please add docstrings for this class.

tothtamas28 commented on pyk/src/pyk/ktool/kfuzz.py:
```diff
@@ -20,13 +21,34 @@
+    def handle_failure_case(self, args: Mapping[EVar, Pattern]) -> None: ...
+
+
+class _KFuzzDefaultTestCaseHandler(KFuzzTestCaseHandler):
```
What do you think about `NullTestCaseHandler` (as a subclass of `KFuzzTestCaseHandler` or `TestCaseHandler`) or `KFuzzNullHandler` (as a subclass of `KFuzzHandler`).

tothtamas28 commented on pyk/src/pyk/ktool/kfuzz.py:
```diff
@@ -20,13 +21,34 @@
 
+class KFuzzTestCaseHandler(ABC):
+    @abstractmethod
+    def handle_test_case(self, args: Mapping[EVar, Pattern]) -> None: ...
```
`handle_test`?

tothtamas28 commented on pyk/src/pyk/ktool/kfuzz.py:
```diff
@@ -20,13 +21,34 @@
+    def handle_test_case(self, args: Mapping[EVar, Pattern]) -> None: ...
+
+    @abstractmethod
+    def handle_failure_case(self, args: Mapping[EVar, Pattern]) -> None: ...
```
`handle_failure`?

gtrepta committed:
Renamings

gtrepta committed:
Docstrings

rv-jenkins committed:
Merge branch 'develop' into pyk/kfuzz-printing
